backend/recommendation.js

```diff
@@ -98,11 +99,60 @@ const createPersonalizedNews = async (req, news, rankings) => {
   }
 };
 
+const changeEnagagementWeights = async (req) => {
+  const interactions = await prisma.userInteraction.findMany({ // all interactions
+    where : {userId : req.session.userId}
+  });
+
+  const prevWeights = await prisma.engagementWeight.findMany({
+    where : {userId : req.session.userId}
+  })
+
+  if (prevWeights[0].updatedLast.toDateString() === (new Date()).toDateString()) { // only want it to update daily
+    return; // not updating
+  }
+
+  // aggregate count, then update the weightage based on the proportion of engagament comes from that specific engagement
+  let openCount = 0;
+  let readCount = 0;
+  let likeCount = 0;
+  let voteCount = 0;
+  for (const interaction of interactions) {
+    openCount += interaction.openCount;
+    readCount += interaction.readCount;
+    likeCount = (interaction.isLiked) ? likeCount + 1 : likeCount;
+    voteCount = (interaction.voted) ? voteCount + 1 : voteCount;
+  }
+
+  const totalCount = openCount + readCount + likeCount + voteCount;
+
+  const weights = {
+    openCount : openCount / totalCount, // proportion of engagement done by this signal
+    readCount : readCount / totalCount,
+    isLiked : likeCount / totalCount,
+    voted : voteCount / totalCount,
+  } 
+
+  for (const signalWeight of prevWeights) {
+    const signal = signalWeight.signal;
+
+    const updatedWeight = await prisma.engagementWeight.update({
+      where : {id : signalWeight.id},
+      data : {
+        weight : (signalWeight.weight + weights[signal]).toFixed(2),
+        updatedLast : new Date()
+      }
+    })
+  }
+}
+
 const calculateEngagement = async (req) => {
   const interactions = await prisma.userInteraction.findMany({
     where: { userId: req.session.userId },
   });
 
+  await changeEnagagementWeights(req);
```

Comment:
I think we can be more efficient here. Right now if i'm understanding correctly we're re-calculating the engagement weights upon every single engagement score calculation? I'm thinking we can update and store the newest weights on each individual user interaction. We can store the total number of each user interactions in EngagementWeights and simply update that alongside updating the interaction node. Then over here all we need to do is fetch the engagementWeight node and some some simple calculations without any looping necessary